pkg/sql/opt/memo/logical_props_builder.go

```diff
@@ -79,10 +85,15 @@ func (b *logicalPropsBuilder) buildScanProps(scan *ScanExpr, rel *props.Relation
 	// ----------------
 	// Initialize not-NULL columns from the table schema.
 	rel.NotNullCols = tableNotNullCols(md, scan.Table)
+	// Union not-NULL columns with not-NULL columns in the constraint.
 	if scan.Constraint != nil {
 		rel.NotNullCols.UnionWith(scan.Constraint.ExtractNotNullCols(b.evalCtx))
 	}
-	rel.NotNullCols.IntersectionWith(rel.OutputCols)
```

Comment:
check_expr.go tests assertions are failing the tests because I've removed this line, causing the set of not-null cols to contain columns not in the RelExpr's output cols.

```
------- Stdout: -------
=== RUN   TestLogicalProps/lookup-join
--- FAIL: TestLogicalProps/lookup-join (0.00s)
opt_tester.go:454: testdata/logprops/lookup-join:24: not null cols (3) not a subset of output cols (1,2)
(1) assertion failure
Wraps: (2) attached stack trace
| github.com/cockroachdb/cockroach/pkg/sql/opt/props.(*Relational).Verify
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/props/verify.go:53
| github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).CheckExpr
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:38
| github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).MemoizeScan
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:16838
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructScan
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:249
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).pruneScanCols
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/prune_cols_funcs.go:331
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).PruneCols
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/prune_cols_funcs.go:249
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1432
```

This is an interesting case because this appears to be the first time that columns can be guaranteed to be not null even though they aren't in the set out output cols. For example, if the partial index `INDEX (a) WHERE b > 0` is used for a scan, `b` is guaranteed to be non-null for any row returned by the scan, even though `b` is not contained in the index (hence not an output col).

If I add this line back in, it alters the stats for partial index scans such that they are slightly off. Notice in the example below the scan estimates ~25.1 rows while the index-join (originally a Select in that memo group) estimates ~24.8 rows. 

```
exec-ddl
CREATE INDEX idx ON a (i) WHERE s IN ('foo', 'bar', 'baz')
----

opt
SELECT * FROM a WHERE i > 10 AND i < 20 AND s IN ('foo', 'bar', 'baz')
----
index-join a
 ├── columns: k:1(int!null) i:2(int!null) s:3(string!null) t:4(string)
 ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── scan a@idx,partial
      ├── columns: k:1(int!null) i:2(int!null)
      ├── constraint: /2/1: [/11 - /19]
      ├── stats: [rows=25.1421429, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=25.1421429, distinct(2,3)=25.1421429, null(2,3)=0]
      ├── key: (1)
      └── fd: (1)-->(2)
```

Without this line, the stats seem more accurate:

```
exec-ddl
CREATE INDEX idx ON a (i) WHERE s IN ('foo', 'bar', 'baz')
----

opt
SELECT * FROM a WHERE i > 10 AND i < 20 AND s IN ('foo', 'bar', 'baz')
----
index-join a
 ├── columns: k:1(int!null) i:2(int!null) s:3(string!null) t:4(string)
 ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── scan a@idx,partial
      ├── columns: k:1(int!null) i:2(int!null)
      ├── constraint: /2/1: [/11 - /19]
      ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
      ├── key: (1)
      └── fd: (1)-->(2)
```